File: m6-expanding-our-first-controller-end/src/main/java/com/baeldung/ls/web/controller/ProjectController.java
package com.baeldung.ls.web.controller;

import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.server.ResponseStatusException;

import com.baeldung.ls.persistence.model.Project;
import com.baeldung.ls.service.IProjectService;

@RestController
@RequestMapping(value = "/projects")
public class ProjectController {

    private IProjectService projectService;

    public ProjectController(IProjectService projectService) {
        this.projectService = projectService;
    }

    @GetMapping(value = "/{id}")
    public Project findOne(@PathVariable Long id) {
        return projectService.findById(id)
            .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND));
    }

    @PostMapping
    public void create(@RequestBody Project project) {
        this.projectService.save(project);
    }

}

File: decouple-controller-from-entities-end-kotlin/src/main/kotlin/com/baeldung/ls/web/ProjectController.kt
package com.baeldung.ls.web

import com.baeldung.ls.persistence.Project
import com.baeldung.ls.persistence.Task
import com.baeldung.ls.service.IProjectService
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.*
import org.springframework.web.server.ResponseStatusException

@RestController
@RequestMapping(value = ["/projects"])
class ProjectController(private val projectService: IProjectService) {

    @GetMapping(value = ["/{id}"])
    fun findOne(@PathVariable id: Long): ProjectDto =
        projectService.findById(id)?.convertToDto() ?: throw ResponseStatusException(HttpStatus.NOT_FOUND)

    @PostMapping
    fun create(@RequestBody newProject: ProjectDto) {
        projectService.save(newProject.convertToEntity())
    }
}

private fun Project.convertToDto(): ProjectDto =
    ProjectDto(id, name, dateCreated).also { projectDto ->
        projectDto.tasks = tasks.map { it.convertToDto() }.toMutableSet()
    }

private fun Task.convertToDto(): TaskDto = TaskDto(id, name, description, dateCreated, dueDate, status)

private fun ProjectDto.convertToEntity(): Project = Project(id, name, dateCreated)

